notebooks/ch16/classification_sgd.md

```diff
@@ -0,0 +1,187 @@
+$$
+
+In logistic regression, we use the cross entropy cost as our cost function. $\nabla_\hat\theta L(\hat\theta, X, y)$ is then the gradient of the cross entropy cost, $-\frac{1}{n}\sum_{i=1}^n(y_i - \sigma_i)X_i$; plugging this in allows us to find the gradient descent algorithm specific to logistic regression. Letting $ \sigma_i = f_\hat\theta(X_i) = \sigma(X_i \cdot \hat \theta) $, this becomes:
+
+$$
+\begin{align}
+\hat\theta_{t+1} &= \hat\theta_t - \alpha \cdot \left(- \frac{1}{n} \sum_{i=1}^{n} \left(y_i - \sigma_i\right) X_i \right) \\
+&= \hat\theta_t + \alpha \cdot \left(\frac{1}{n} \sum_{i=1}^{n} \left(y_i - \sigma_i\right) X_i \right)
+\end{align}
+$$
+
+### Stochastic Gradient Descent
+
+The general update formula is below, where $l(\hat\theta, X, y)$ is the loss function for a single data point:
+
+$$
+\hat\theta_{t+1} = \hat\theta_t - \alpha \nabla_\hat\theta l(\hat\theta, X, y)
+$$
+
+Returning back to our example in logistic regression, we approximate the gradient of the cross entropy cost using the gradient of the cross entropy loss of one data point. This is shown below, with $ \sigma_i = f_\hat{\theta}(X_i) = \sigma(X_i \cdot \hat \theta) $.
+
+$$
+\begin{align}
+\nabla_\hat\theta L(\hat\theta, X, y) &\approx \nabla_\hat\theta l(\hat\theta, X, y)\\
+&= -(y_i - \sigma_i)X_i
+\end{align}
+$$
+
+When we plug this approximation into the general formula for stochastic gradient descent, we find the stochastic gradient descent update formula for logistic regression.
+
+$$
+\begin{align}
+\hat\theta_{t+1} &= \hat\theta_t - \alpha \nabla_\hat\theta l(\hat\theta, X, y) \\
+&= \hat\theta_t + \alpha \cdot (y_i - \sigma_i)X_i
+\end{align}
+$$
+
+### Mini-batch Gradient Descent
+
+Similarly, we can approximate the gradient of the cross entropy cost using a random sample of data points, also known as a mini-batch.
+
+$$
+\nabla_\hat\theta L(\hat\theta, X, y) \approx \frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}\nabla_{\hat\theta}l(\hat\theta, X, y_i)
+$$
+
+We substitute this for the gradient of the cross entropy cost, yielding a mini-batch gradient descent update formula specific to logistic regression:
+
+$$
+\begin{align}
+\hat\theta_{t+1} &= \hat\theta_t - \alpha \cdot -\frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}(y_i - \sigma_i)X_i \\
+&= \hat\theta_t + \alpha \cdot \frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}(y_i - \sigma_i)X_i
+\end{align}
+$$
+
+## Implementation in Scikit-Learn
+
+Scikit-learn has implementations for stochastic gradient descent and mini-batch gradient descent using the [`SGDClassifier`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) class. To gain a better estimate of the speed benefits afforded by stochastic gradient descent and mini-batch gradient descent, we will first manually implement batch gradient descent; then, we will compare the results with stochastic gradient descent and mini-batch gradient descent.
```

Comment:
Include a short one-two line description of what the code boxes are doing, along the lines of "the code below trains a classifier predicting LeBron's shots using the three GD algorithms, and you can see that ____ is the fastest."